// mediafire/dist/mediafire.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var User_1 = require("./api/User");
var File_1 = require("./api/File");
var Folder_1 = require("./api/Folder");
var Device_1 = require("./api/Device");
var System_1 = require("./api/System");
var MediaFire = /** @class */ (function () {
    function MediaFire(host, user, file, folder, device, system) {
        if (user === void 0) { user = new User_1.User(host); }
        if (file === void 0) { file = new File_1.File(host); }
        if (folder === void 0) { folder = new Folder_1.Folder(host); }
        if (device === void 0) { device = new Device_1.Device(host); }
        if (system === void 0) { system = new System_1.System(host); }
        this.host = host;
        this.user = user;
        this.file = file;
        this.folder = folder;
        this.device = device;
        this.system = system;
    }
    MediaFire.prototype.authenticate = function (token) {
        this.user = new User_1.User(this.host, token);
        this.file = new File_1.File(this.host, token);
        this.folder = new Folder_1.Folder(this.host, token);
        this.device = new Device_1.Device(this.host, token);
    };
    return MediaFire;
}());
exports.default = MediaFire;
